README.md
# georef-ar-py
Paquete para consultas y procesamiento de respuestas sobre la API de georef-ar

## Installation

### Install from PyPi:

`pip install georef-ar-py`

Para usar desde la línea de comandos:

`pip install "georef-ar-py[cli]"`


## Ejemplo de uso por linea de comando:

`geoarpy --help`
```
Usage: geoarpy [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  batch-normalize  geoarpy batch normalize Commandline
  diff             geoarpy diff Commandline
  info             geoarpy info Commandline
  normalize        geoarpy normalize Commandline
```

### Obtener un resumen de una API
`geoarpy info --help`
```
Usage: geoarpy info [OPTIONS]

  geoarpy info Commandline

  Obtienen un resumen de los registros en la API especificada

  url es el path a la API destino que se quiere consultar.

Options:
  --url TEXT    [default: https://apis.datos.gob.ar/georef/api/]
  --token TEXT  Un token para enviar en el encabezado de cada petición
  --debug
  --help        Show this message and exit.
```

Obtener los datos de una API propia

`geoarpy info --url "http://mi-api.ar/georef/api/"`

Obtener los datos de la API oficial (https://apis.datos.gob.ar/georef/api/)

`geoarpy info`

### Obtener las diferencias entre dos API's
`geoarpy diff --help`
```
Usage: geoarpy diff [OPTIONS] URL

  geoarpy diff Commandline

  Calcula las diferencias de registros entre dos API's. Guarda el resultado en
  archivos de formato csv y/o json.

  url es el path a la API destino que se quiere comparar.

Options:
  --origin_url TEXT               [default:
                                  https://apis.datos.gob.ar/georef/api/]
  --token TEXT                    Un token para enviar en el encabezado de
                                  cada petición
  --layer [provincias|departamentos|municipios|localidades-censales|asentamientos|localidades|calles]
  --extension [json|csv|both]     [default: both]
  --debug
  --help                          Show this message and exit.
```

Comparar los datos de una API propia contra la API oficial

`geoarpy diff "http://mi-api.ar/georef/api/"`

Comparar los datos entre dos API's

`geoarpy diff "http://mi-api_1.ar/georef/api/" --origin-url "http://mi-api_2.ar/georef/api/"`

### Normalizar una dirección (Obtener la nomenclatura)
`geoarpy normalize --help`
```
Usage: geoarpy normalize [OPTIONS] ADDRESS

  geoarpy normalize Commandline

  Dada una direccion suministra la misma normalizada (nomenclatura) o no
  devuelve nada si no se pudo normalizar

  url es el path a la API destino que se quiere consultar.

Options:
  --url TEXT               [default: https://apis.datos.gob.ar/georef/api/]
  --token TEXT             Un token para enviar en el encabezado de cada
                           petición
  --provincia TEXT
  --departamento TEXT
  --localidad_censal TEXT
  --localidad TEXT
  --debug
  --help                   Show this message and exit.
```

Normaliza una dirección indicando la localidad

`geoarpy normalize "feliz san marti 390" --localidad_censal="junin del los andes"`

Output:

`FELIX SAN MARTIN 390, Huiliches, Neuquén`

### Normalizar un conjunto de direcciones
`geoarpy batch-normalize --help`
```
Usage: geoarpy batch-normalize [OPTIONS] INPUT_CSV OUTPUT_CSV

  geoarpy batch normalize Commandline

  Normaliza un archivo csv con direcciones.  Se suministra un archivo
  (input_csv) y se leen las direcciones bajo el encabezado "direccion".
  Optativamente, se pueden suministrar columnas con información extra bajo los
  siguientes encabezados:     "localidad_censal", "localidad", "departamento",
  "provincia"

  Escribe los resultados a un archivo csv (output_csv)

Options:
  --url TEXT            [default: https://apis.datos.gob.ar/georef/api/]
  --token TEXT          Un token para enviar en el encabezado de cada petición
  --chunk_size INTEGER  Cantidad de registros a leer desde un csv en cada
                        procesamiento  [default: 1000]
  --data_size INTEGER   Cantidad de registros a enviar en cada petición POST
                        [default: 500]
  --rps INTEGER         Número máximo de peticiones por segundo
  --debug
  --help                Show this message and exit.

```

Lee las direcciones de un archivo csv y escribe los datos normalizados en un nuevo archivo

`geoarpy batch-normalize ./direciones.csv ./direcciones_normalizadas.csv`


CHANGELOG.md
# Changelog

Todos los cambios notables del proyecto serán documentados en este archivo

El formato está basado en [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
y este proyecto adhiere a [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

## [0.0.6] - 2023-09-23

### Added

- Implementa consultas asincronas en la normalización de direcciones por lotes para disminuir el tiempo
- Lectura de csv por chunks para controla la memoria 
- Agrega un ratelimiter rudimentario para la normalización de direcciones

## [0.0.5] - 2023-09-15

### Added

- Implementa la normalización de direcciones, simple y por lote

## [0.0.4] - 2023-08-10

### Added

- Implementa asincronismo en las consultas
- Implementa multiprocesamiento en la generación de informes

## [0.0.3] - 2023-06-19

### Added

- Implementa token en headers
- Corrige argumento en la línea de comandos
- Corrige bugs en la consulta por regiones y otros
- Modularización de algunas funciones

## [0.0.2] - 2023-06-18

### Added

- Archivo CHANGELOG.
- Implementación de comandos de cli
- Funciones para comparación de dos API's y generación de reportes
- Implementación de generación de resumen de la API
- Implementación de logs

## [0.0.1] - 2023-04-28

### Added

- Dependencia a requests==2.29.
- Archivo README y LICENSE.
- Tres funciones de consulta a la API tomadas de la documentación.

[unreleased]: https://github.com/pavloae/georef-ar-py
[0.0.6]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.6
[0.0.5]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.5
[0.0.4]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.4
[0.0.3]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.3
[0.0.2]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.2
[0.0.1]: https://github.com/pavloae/georef-ar-py/releases/tag/0.0.1

src/georef_ar_py/commands.py
import asyncio
import concurrent.futures
import json
import logging
import os

import click

from . import georequests
from .constants import ENTITIES
from .diff import process
from .georequests import API_BASE_URL
from .info import get_resume
from .normalization import AddressNormalizer, Address


def get_logger(level):
    logging.basicConfig(level=level, format="%(levelname)s - %(message)s")
    logger = logging.getLogger(__name__)
    return logger


@click.group()
def cli():
    pass


@cli.command()
@click.argument('url', type=str)
@click.option('--origin_url', required=False, type=str, show_default=True, default=API_BASE_URL)
@click.option(
    '--token', required=False, type=str, show_default=True, default=None,
    help="Un token para enviar en el encabezado de cada petición"
)
@click.option(
    '--layer', required=False, type=click.Choice(ENTITIES, case_sensitive=True),
    show_default=True, default=None
)
@click.option(
    '--extension', required=False, type=click.Choice(['json', 'csv', 'both'], case_sensitive=True),
    show_default=True, default="both"
)
@click.option('--debug', is_flag=True, show_default=False)
def diff(*args, **kwargs):
    """
    geoarpy diff Commandline

    Calcula las diferencias de registros entre dos API's.
    Guarda el resultado en archivos de formato csv y/o json.

    url es el path a la API destino que se quiere comparar.
    """
    debug = kwargs.pop('debug')
    log = get_logger(logging.DEBUG if debug else logging.INFO)
    log.info("Comenzando la generación de diferencias...")

    src_url = kwargs.pop('origin_url')
    target_url = kwargs.pop('url')
    ext = kwargs.pop('extension')
    layer = kwargs.pop('layer')
    georequests.__dict__['TOKEN'] = kwargs.pop('token')

    path_dir = os.getcwd()

    if ext and ext not in ['json', 'csv', 'both']:
        raise NotImplemented(f'La extensión {ext} no está implementada.')

    entities = layer if layer else ENTITIES

    if isinstance(entities, str):
        entities = [layer]

    with concurrent.futures.ProcessPoolExecutor() as executor:
        for entity in entities:
            executor.submit(process, src_url, target_url, entity, path_dir, ext)


@cli.command()
@click.option('--url', required=False, type=str, show_default=True, default=API_BASE_URL)
@click.option(
    '--token', required=False, type=str, show_default=True, default=None,
    help="Un token para enviar en el encabezado de cada petición"
)
@click.option('--debug', is_flag=True, show_default=False)
def info(*args, **kwargs):
    """
    geoarpy info Commandline

    Obtienen un resumen de los registros en la API especificada

    url es el path a la API destino que se quiere consultar.
    """
    debug = kwargs.pop('debug')
    log = get_logger(logging.DEBUG if debug else logging.INFO)
    log.info("Comenzando la generación del resumen...")

    target_url = kwargs.pop('url')
    georequests.__dict__['TOKEN'] = kwargs.pop('token')

    resume = asyncio.run(get_resume(target_url))

    filename = os.path.join(os.getcwd(), 'info.json')
    log.info(f"Guardando archivo en {filename}")
    with open(filename, '+w') as f:
        json.dump(resume, f)


@cli.command()
@click.argument('direccion', type=str)
@click.option('--url', required=False, type=str, show_default=True, default=API_BASE_URL)
@click.option(
    '--token', required=False, type=str, show_default=True, default=None,
    help="Un token para enviar en el encabezado de cada petición"
)
@click.option('--provincia', required=False, type=str, show_default=True, default=None)
@click.option('--departamento', required=False, type=str, show_default=True, default=None)
@click.option('--localidad_censal', required=False, type=str, show_default=True, default=None)
@click.option('--localidad', required=False, type=str, show_default=True, default=None)
@click.option('--debug', is_flag=True, show_default=False)
def normalize(*args, **kwargs):
    """
    geoarpy normalize Commandline

    Dada una direccion suministra la misma normalizada (nomenclatura) o no devuelve nada si no se pudo normalizar

    url es el path a la API destino que se quiere consultar.
    """
    debug = kwargs.pop('debug')
    get_logger(logging.DEBUG if debug else logging.INFO)

    target_url = kwargs.pop('url')
    georequests.__dict__['TOKEN'] = kwargs.pop('token')

    address = Address(**kwargs)

    address_normalizer = AddressNormalizer(target_url)

    address_normalizer.normalize(address)

    print(address.nomenclature)


@cli.command()
@click.argument('input_csv', type=click.Path(exists=True))
@click.argument('output_csv', type=click.Path(writable=True))
@click.option('--url', required=False, type=str, show_default=True, default=API_BASE_URL)
@click.option(
    '--token', required=False, type=str, show_default=True, default=None,
    help="Un token para enviar en el encabezado de cada petición"
)
@click.option(
    '--chunk_size', required=False, type=int, show_default=True, default=1000,
    help="Cantidad de registros a leer desde un csv en cada procesamiento"
)
@click.option(
    '--data_size', required=False, type=int, show_default=True, default=500,
    help="Cantidad de registros a enviar en cada petición POST"
)
@click.option(
    '--rps', required=False, type=int, show_default=True, default=None,
    help="Número máximo de peticiones por segundo"
)
@click.option('--debug', is_flag=True, show_default=False)
def batch_normalize(input_csv, output_csv, **kwargs):
    """
    geoarpy batch normalize Commandline

    Normaliza un archivo csv con direcciones.
     Se suministra un archivo (input_csv) y se leen las direcciones bajo el encabezado "direccion".
     Optativamente, se pueden suministrar columnas con información extra bajo los siguientes encabezados:
        "localidad_censal", "localidad", "departamento", "provincia"

    Escribe los resultados a un archivo csv (output_csv)
    """

    debug = kwargs.pop('debug')
    log = get_logger(logging.DEBUG if debug else logging.INFO)
    log.info("Comenzando la normalización...")

    target_url = kwargs.pop('url')
    georequests.__dict__['TOKEN'] = kwargs.pop('token')

    normalizer = AddressNormalizer(
        url=target_url,
        chunk_size=kwargs.pop('chunk_size'),
        data_size=kwargs.pop('data_size'),
        rps=kwargs.pop('rps')
    )

    normalizer.csv2csv(input_csv, output_csv)
